dawn/main/ansible/Ansible.js

```diff
@@ -117,62 +44,212 @@ function buildProto(data) {
     });
     return GamepadMsg;
   });
+
   const message = new DawnData({
     student_code_status: status,
     gamepads,
   });
+
   return message;
 }
 
-/*
- * IPC Connection with sagas.js' ansibleGamepads()
- * Sends messages when Gamepad information changes
- * or when 100 ms has passed (with 50 ms cooldown)
- */
-ipcMain.on('stateUpdate', (event, data) => {
-  const message = buildProto(data).encode().toBuffer();
-  console.log('Dawn Sent UDP');
-  sendSocket.send(message, sendPort, runtimeIP, (err) => {
-    if (err) {
-      console.error('UDP Sending Error');
+const LISTEN_PORT = 1235;
+class ListenSocket {
+  constructor() {
+    // TODO: Verify if reuseAddr prevents EADDRINUSE
+    this.socket = dgram.createSocket({ type: 'udp4', reuseAddr: true });
+
+    /*
+     * Runtime message handler. Sends robot state to store.info
+     * and raw sensor array to peripheral reducer
+     */
+    this.socket.on('message', (msg) => {
+      try {
+        const { robot_state, sensor_data } = RuntimeData.decode(msg);
+        console.log('Dawn received UDP');
+        RendererBridge.reduxDispatch(infoPerMessage(robot_state));
+        RendererBridge.reduxDispatch(updatePeripherals(sensor_data));
+      } catch (err) {
+        console.log('Error decoding UDP');
+        throw err;
+      }
+    });
+
+    this.socket.on('error', (err) => {
+      console.log('UDP listening error');
+      this.socket.close();
+      throw err;
+    });
+
+    this.socket.on('close', () => {
+      RendererBridge.reduxDispatch(ansibleDisconnect());
+      console.log('UDP listening closed');
+    });
+
+    this.socket.bind(LISTEN_PORT);
+  }
+
+  close() {
```

Comment:
It's not, but I wasn't sure how to avoid that code duplication, since the sockets are not always the same (between the UDP and the TCP sockets), so the methods you call for closing aren't the same (`.close` vs `.end`).